pkg/controller/traceflow/controller_test.go

```diff
@@ -0,0 +1,146 @@
+
+package traceflow
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	k8stesting "k8s.io/client-go/testing"
+
+	ops "github.com/vmware-tanzu/antrea/pkg/apis/ops/v1alpha1"
+	"github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+	fakeversioned "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned/fake"
+	crdinformers "github.com/vmware-tanzu/antrea/pkg/client/informers/externalversions"
+)
+
+var alwaysReady = func() bool { return true }
+
+const informerDefaultResync time.Duration = 30 * time.Second
+
+type traceflowController struct {
+	*Controller
+	client             versioned.Interface
+	informerFactory    informers.SharedInformerFactory
+	crdInformerFactory crdinformers.SharedInformerFactory
+}
+
+func newController() *traceflowController {
+	client := fake.NewSimpleClientset()
+	crdClient := newCRDClientset()
+	informerFactory := informers.NewSharedInformerFactory(client, informerDefaultResync)
+	crdInformerFactory := crdinformers.NewSharedInformerFactory(crdClient, informerDefaultResync)
+	controller := NewTraceflowController(crdClient,
+		informerFactory.Core().V1().Pods(),
+		crdInformerFactory.Ops().V1alpha1().Traceflows())
+	controller.traceflowListerSynced = alwaysReady
+	return &traceflowController{
+		controller,
+		crdClient,
+		informerFactory,
+		crdInformerFactory,
+	}
+}
+
+func TestTraceflow(t *testing.T) {
+	// Use shorter timeout.
+	timeoutDuration = 2 * time.Second
+	timeoutCheckInterval = timeoutDuration / 2
+
+	tfc := newController()
+	stopCh := make(chan struct{})
+	tfc.crdInformerFactory.Start(stopCh)
+	go tfc.Run(stopCh)
+
+	tf1 := ops.Traceflow{
+		ObjectMeta: metav1.ObjectMeta{Name: "tf1", UID: "uid1"},
+		Spec: ops.TraceflowSpec{
+			Source:      ops.Source{"ns1", "pod1"},
+			Destination: ops.Destination{Namespace: "ns2", Pod: "pod2"},
+		},
+	}
+
+	tfc.client.OpsV1alpha1().Traceflows().Create(context.TODO(), &tf1, metav1.CreateOptions{})
+	time.Sleep(1 * time.Second)
+	res, _ := tfc.client.OpsV1alpha1().Traceflows().Get(context.TODO(), "tf1", metav1.GetOptions{})
+	assert.Equal(t, res.Status.Phase, ops.Running)
+	// DataplaneTag should be allocated by Controller.
+	assert.True(t, res.Status.DataplaneTag > 0)
+	tfc.runningTraceflowsMutex.Lock()
+	assert.Equal(t, len(tfc.runningTraceflows), 1)
+	tfc.runningTraceflowsMutex.Unlock()
+
+	// Test Controller handling of succeeded Traceflow.
+	res.Status.Results = []ops.NodeResult{
+		// Sender
+		ops.NodeResult{
+			Observations: []ops.Observation{ops.Observation{Component: ops.SpoofGuard}},
+		},
+		// Receiver
+		ops.NodeResult{
+			Observations: []ops.Observation{ops.Observation{Action: ops.Delivered}},
+		},
+	}
+	tfc.client.OpsV1alpha1().Traceflows().Update(context.TODO(), res, metav1.UpdateOptions{})
+	time.Sleep(1 * time.Second)
```

Comment:
it's a small `Sleep` but I think a call to `wait.Poll` (or one of the variants in that package) would be better, especially if it usually takes a few 10s of ms for the CRD status to change

same below (even for `time.Sleep(timeoutDuration)`)

Reply:
Done.